File: cmd/task.go
package cmd

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/Pernek-Enterprises/dispatch/internal/config"
	"github.com/Pernek-Enterprises/dispatch/internal/jobs"
	"github.com/Pernek-Enterprises/dispatch/internal/log"
	"github.com/Pernek-Enterprises/dispatch/internal/pipe"
	"github.com/Pernek-Enterprises/dispatch/internal/workflows"
)

func Task(args []string) {
	if len(args) == 0 {
		fmt.Println(`dispatch task — manage tasks

Commands:
  dispatch task create "description" --workflow coding-easy
  dispatch task create --interactive
  dispatch task list
  dispatch task show <task-id>`)
		return
	}

	switch args[0] {
	case "create":
		taskCreate(args[1:])
	case "list":
		taskList()
	case "show":
		if len(args) < 2 {
			fmt.Fprintln(os.Stderr, "Usage: dispatch task show <task-id>")
			os.Exit(1)
		}
		taskShow(args[1])
	default:
		fmt.Fprintf(os.Stderr, "Unknown task command: %s\n", args[0])
		os.Exit(1)
	}
}

func taskCreate(args []string) {
	var description string
	var workflowName string
	var priority string
	interactive := false

	// Parse flags (flags can appear anywhere)
	var descParts []string
	i := 0
	for i < len(args) {
		switch args[i] {
		case "--workflow", "-w":
			i++
			if i < len(args) {
				workflowName = args[i]
			}
		case "--priority", "-p":
			i++
			if i < len(args) {
				priority = args[i]
			}
		case "--interactive", "-i":
			interactive = true
		default:
			descParts = append(descParts, args[i])
		}
		i++
	}
	description = strings.Join(descParts, " ")

	if interactive {
		taskCreateInteractive()
		return
	}

	if description == "" {
		fmt.Fprintln(os.Stderr, "Usage: dispatch task create \"description\" --workflow coding-easy")
		os.Exit(1)
	}

	// Default workflow
	if workflowName == "" {
		// List available and pick first, or error
		available, _ := workflows.ListAll()
		if len(available) == 1 {
			workflowName = available[0]
		} else {
			fmt.Fprintln(os.Stderr, "Multiple workflows available. Specify one with --workflow:")
			for _, w := range available {
				fmt.Fprintf(os.Stderr, "  %s\n", w)
			}
			os.Exit(1)
		}
	}

	if priority == "" {
		priority = "normal"
	}

	taskID := createTask(description, workflowName, priority)
	fmt.Printf("✓ Task created: %s\n  workflow: %s\n  priority: %s\n", taskID, workflowName, priority)
}

func taskCreateInteractive() {
	reader := bufio.NewReader(os.Stdin)
	ask := func(prompt string) string {
		fmt.Print(prompt)
		line, _ := reader.ReadString('\n')
		return strings.TrimSpace(line)
	}

	fmt.Println("\n  Create a new task\n")

	description := ask("  Description: ")
	if description == "" {
		fmt.Fprintln(os.Stderr, "Description required")
		os.Exit(1)
	}

	// Show available workflows
	available, _ := workflows.ListAll()
	fmt.Println("\n  Available workflows:")
	for _, w := range available {
		wf, _ := workflows.Load(w)
		desc := ""
		if wf != nil {
			desc = " — " + wf.Description
		}
		fmt.Printf("    %s%s\n", w, desc)
	}
	workflowName := ask("\n  Workflow: ")
	if workflowName == "" && len(available) == 1 {
		workflowName = available[0]
	}
	if workflowName == "" {
		fmt.Fprintln(os.Stderr, "Workflow required")
		os.Exit(1)
	}

	priority := ask("  Priority (normal/high/urgent) [normal]: ")
	if priority == "" {
		priority = "normal"
	}

	// Optional: multi-line context
	fmt.Println("  Additional context (empty line to finish):")
	var contextLines []string
	for {
		line := ask("    ")
		if line == "" {
			break
		}
		contextLines = append(contextLines, line)
	}

	fullDescription := description
	if len(contextLines) > 0 {
		fullDescription += "\n\n## Context\n" + strings.Join(contextLines, "\n")
	}

	taskID := createTask(fullDescription, workflowName, priority)
	fmt.Printf("\n  ✓ Task created: %s\n    workflow: %s\n    priority: %s\n\n", taskID, workflowName, priority)
}

func createTask(description, workflowName, priority string) string {
	// Validate workflow exists
	wf, err := workflows.Load(workflowName)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading workflow %q: %v\n", workflowName, err)
		os.Exit(1)
	}

	// Get first step
	firstStep, ok := wf.Steps[wf.FirstStep]
	if !ok {
		fmt.Fprintf(os.Stderr, "First step %q not found in workflow\n", wf.FirstStep)
		os.Exit(1)
	}

	// Generate task ID
	taskID := jobs.NewTaskID()

	// Build the prompt for the first step
	prompt := fmt.Sprintf("# Task: %s\n\n%s\n\n## Step: %s\n\n%s",
		taskID, description, wf.FirstStep, firstStep.Prompt)

	artifactDir := filepath.Join(config.Root, "artifacts")

	jobType := "work"
	if firstStep.Agent == "stefan" || firstStep.Type == "human" {
		jobType = "human"
	}

	// Create the first job
	_, err = jobs.Create(jobs.CreateOpts{
		Task:     taskID,
		Workflow: workflowName,
		Step:     wf.FirstStep,
		Agent:    firstStep.Agent,
		Model:    firstStep.Model,
		Type:     jobType,
		Priority: priority,
		Timeout:  firstStep.Timeout,
		Prompt:   prompt,
	})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating job: %v\n", err)
		os.Exit(1)
	}

	// Create artifacts dir for this task
	os.MkdirAll(filepath.Join(artifactDir, taskID), 0755)

	// Save the task description as an artifact
	os.WriteFile(
		filepath.Join(artifactDir, taskID, "task.md"),
		[]byte(description+"\n"),
		0644,
	)

	// Notify foreman if pipe exists
	cfg, _ := config.Load()
	pipePath := "/tmp/dispatch.pipe"
	if cfg != nil && cfg.PipePath != "" {
		pipePath = cfg.PipePath
	}
	if _, err := os.Stat(pipePath); err == nil {
		pipe.Send(pipePath, pipe.Message{
			Type:   "new_task",
			TaskID: taskID,
		})
	}

	log.Info("Task created: %s (workflow=%s)", taskID, workflowName)
	return taskID
}

func taskList() {
	// Show tasks from state.json + scan job folders
	statePath := filepath.Join(config.Root, "state.json")
	data, _ := os.ReadFile(statePath)

	var s struct {
		Tasks map[string]struct {
			Workflow    string `json:"workflow"`
			CurrentStep string `json:"currentStep"`
			Status     string `json:"status"`
		} `json:"tasks"`
	}
	json.Unmarshal(data, &s)

	// Also scan pending/active for tasks not yet in state
	pendingJobs, _ := jobs.List("pending")
	activeJobs, _ := jobs.List("active")
	allJobs := append(pendingJobs, activeJobs...)

	// Collect unique tasks
	tasks := make(map[string]string) // taskID → status
	for id, t := range s.Tasks {
		tasks[id] = fmt.Sprintf("%s (%s/%s)", t.Status, t.Workflow, t.CurrentStep)
	}
	for _, j := range allJobs {
		if j.Task != "" {
			if _, exists := tasks[j.Task]; !exists {
				tasks[j.Task] = fmt.Sprintf("pending (%s/%s)", j.Workflow, j.Step)
			}
		}
	}

	if len(tasks) == 0 {
		fmt.Println("No tasks. Create one with: dispatch task create \"description\" --workflow coding-easy")
		return
	}

	fmt.Println("Tasks:")
	for id, status := range tasks {
		fmt.Printf("  %s — %s\n", id, status)
	}
}

func taskShow(taskID string) {
	// Check state
	statePath := filepath.Join(config.Root, "state.json")
	data, _ := os.ReadFile(statePath)

	var s struct {
		Tasks map[string]json.RawMessage `json:"tasks"`
	}
	json.Unmarshal(data, &s)

	if taskData, ok := s.Tasks[taskID]; ok {
		var pretty map[string]interface{}
		json.Unmarshal(taskData, &pretty)
		out, _ := json.MarshalIndent(pretty, "  ", "  ")
		fmt.Printf("Task: %s\n  %s\n", taskID, string(out))
	} else {
		fmt.Printf("Task: %s (not in state)\n", taskID)
	}

	// Show task description
	descPath := filepath.Join(config.Root, "artifacts", taskID, "task.md")
	if desc, err := os.ReadFile(descPath); err == nil {
		fmt.Printf("\nDescription:\n  %s\n", strings.TrimSpace(string(desc)))
	}

	// Show related jobs
	fmt.Println("\nJobs:")
	for _, folder := range []string{"pending", "active", "done", "failed"} {
		jobList, _ := jobs.List(folder)
		for _, j := range jobList {
			if j.Task == taskID || strings.HasPrefix(j.ID, taskID[:8]) {
				fmt.Printf("  [%s] %s — step:%s agent:%s\n", folder, j.ID, j.Step, j.Agent)
			}
		}
	}
}

File: cli/task.ts
import * as fs from "fs";
import * as path from "path";
import * as readline from "readline";
import { ROOT, loadConfig } from "../config.js";
import { createJob, newTaskId, listJobs } from "../jobs.js";
import { loadWorkflow, listWorkflows, getRole, validateWorkflowRoles } from "../workflows.js";
import { sendPipe } from "../pipe.js";
import { loadSystemPromptPublic } from "../prompts.js";
import { loadProject, listProjects, validateProjectHooks } from "../project.js";

export function taskCmd(args: string[]): void {
  if (!args.length || args[0] === "help") {
    console.log(`dispatch task — manage tasks

Commands:
  dispatch task create "description" --workflow coding-easy [--project agentictodo]
  dispatch task create --interactive
  dispatch task list
  dispatch task show <task-id>
  dispatch task projects`);
    return;
  }

  switch (args[0]) {
    case "create": taskCreate(args.slice(1)); break;
    case "list":   taskList(); break;
    case "projects": taskProjects(); break;
    case "show":
      if (args.length < 2) { console.error("Usage: dispatch task show <task-id>"); process.exit(1); }
      taskShow(args[1]);
      break;
    default:
      console.error(`Unknown task command: ${args[0]}`);
      process.exit(1);
  }
}

function taskCreate(args: string[]): void {
  const descParts: string[] = [];
  let workflowName = "";
  let priority = "normal";
  let interactive = false;
  let projectName = "";

  for (let i = 0; i < args.length; i++) {
    if ((args[i] === "--workflow" || args[i] === "-w") && args[i + 1]) { workflowName = args[++i]; }
    else if ((args[i] === "--priority" || args[i] === "-p") && args[i + 1]) { priority = args[++i]; }
    else if ((args[i] === "--project" || args[i] === "-P") && args[i + 1]) { projectName = args[++i]; }
    else if (args[i] === "--interactive" || args[i] === "-i") { interactive = true; }
    else { descParts.push(args[i]); }
  }

  if (interactive) { taskCreateInteractive(); return; }

  const description = descParts.join(" ");
  if (!description) {
    console.error(`Usage: dispatch task create "description" --workflow coding-easy [--project myproject]`);
    process.exit(1);
  }

  if (!workflowName) {
    const available = listWorkflows();
    if (available.length === 1) { workflowName = available[0]; }
    else {
      console.error("Multiple workflows available. Specify one with --workflow:\n" + available.map(w => `  ${w}`).join("\n"));
      process.exit(1);
    }
  }

  // Validate project if specified
  if (projectName) {
    try { loadProject(projectName); } catch (e) {
      console.error(`Project not found: ${projectName}\nAvailable: ${listProjects().join(", ") || "(none)"}`);
      process.exit(1);
    }
  }

  const taskId = doCreateTask(description, workflowName, priority, projectName);
  const projectLine = projectName ? `\n  project:  ${projectName}` : "";
  console.log(`✓ Task created: ${taskId}\n  workflow: ${workflowName}\n  priority: ${priority}${projectLine}`);
}

function taskCreateInteractive(): void {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const ask = (prompt: string): Promise<string> => new Promise(r => rl.question(`  ${prompt}`, (a) => r(a.trim())));

  (async () => {
    console.log("\n  Create a new task\n");
    const description = await ask("Description: ");
    if (!description) { console.error("Description required"); process.exit(1); }

    const available = listWorkflows();
    console.log("\n  Available workflows:");
    for (const w of available) {
      try { const wf = loadWorkflow(w); console.log(`    ${w} — ${wf.description ?? ""}`); }
      catch { console.log(`    ${w}`); }
    }

    let workflowName = await ask("\n  Workflow: ");
    if (!workflowName && available.length === 1) workflowName = available[0];
    if (!workflowName) { console.error("Workflow required"); process.exit(1); }

    let priority = await ask("Priority (normal/high/urgent) [normal]: ");
    if (!priority) priority = "normal";

    console.log("  Additional context (empty line to finish):");
    const contextLines: string[] = [];
    while (true) {
      const line = await ask("    ");
      if (!line) break;
      contextLines.push(line);
    }

    const fullDescription = contextLines.length
      ? `${description}\n\n## Context\n${contextLines.join("\n")}`
      : description;

    const taskId = doCreateTask(fullDescription, workflowName, priority);
    console.log(`\n  ✓ Task created: ${taskId}\n    workflow: ${workflowName}\n    priority: ${priority}\n`);
    rl.close();
  })().catch((e) => { console.error(e); process.exit(1); });
}

function taskProjects(): void {
  const projects = listProjects();
  if (!projects.length) {
    console.log(`No projects. Create one at: ~/.dispatch/projects/<name>.json`);
    return;
  }
  console.log("Projects:");
  for (const name of projects) {
    try {
      const p = loadProject(name);
      const ws = p.workspace ? `  workspace: ${p.workspace}` : "";
      console.log(`  ${name}${p.description ? ` — ${p.description}` : ""}${ws ? `\n${ws}` : ""}`);
    } catch {
      console.log(`  ${name} (failed to load)`);
    }
  }
}

function doCreateTask(description: string, workflowName: string, priority: string, projectName = ""): string {
  const wf = loadWorkflow(workflowName);
  const firstStep = wf.steps[wf.firstStep];
  if (!firstStep) { console.error(`First step "${wf.firstStep}" not found in workflow`); process.exit(1); }

  // Validate all roles in the workflow have corresponding agent files
  const roleWarnings = validateWorkflowRoles(wf);
  if (roleWarnings.length > 0) {
    console.warn(`\n⚠️  Role validation failed for workflow "${workflowName}":`);
    for (const w of roleWarnings) console.warn(`   • ${w}`);
    console.warn(`   Create the missing agent file(s) in ~/.dispatch/agents/ to fix this.\n`);
  }

  // Validate project hooks against workflow steps — warn but don't block
  if (projectName) {
    try {
      const proj = loadProject(projectName);
      const hookWarnings = validateProjectHooks(proj, wf.steps);
      if (hookWarnings.length > 0) {
        console.warn(`\n⚠️  Hook/workflow mismatch for project "${projectName}" + workflow "${workflowName}":`);
        for (const w of hookWarnings) console.warn(`   • ${w}`);
        console.warn(`   Hooks with no matching step will silently not fire.\n`);
      }
    } catch {
      // loadProject failure is handled downstream — skip validation
    }
  }

  const taskId = newTaskId();
  const role = getRole(firstStep);
  const systemPrompt = loadSystemPromptPublic(role);
  const stepPrompt = `# Task: ${taskId}\n\n${description}\n\n## Step: ${wf.firstStep}\n\n${firstStep.prompt ?? ""}`;
  const prompt = systemPrompt ? `${systemPrompt}\n\n---\n\n${stepPrompt}` : stepPrompt;

  const jobType = (firstStep.agent === "stefan" || firstStep.type === "human") ? "human" : "work";

  createJob({
    task: taskId,
    workflow: workflowName,
    step: wf.firstStep,
    agent: role,
    model: firstStep.model,
    type: jobType,
    priority,
    timeout: firstStep.timeout ?? 120,
    project: projectName || undefined,
    loop: firstStep.loop,
    maxLoopIterations: firstStep.maxLoopIterations,
    prompt,
  });

  // Create artifact dir + save task description
  const artifactDir = path.join(ROOT, "artifacts", taskId);
  fs.mkdirSync(artifactDir, { recursive: true });
  fs.writeFileSync(path.join(artifactDir, "task.md"), description + "\n", "utf8");

  // Notify foreman if running
  const cfg = loadConfig();
  if (fs.existsSync(cfg.pipePath)) {
    try { sendPipe(cfg.pipePath, { type: "new_task", taskId }); } catch {}
  }

  return taskId;
}

function taskList(): void {
  const statePath = path.join(ROOT, "state.json");
  let tasks: Record<string, { workflow: string; currentStep: string; status: string }> = {};
  if (fs.existsSync(statePath)) {
    try { tasks = JSON.parse(fs.readFileSync(statePath, "utf8")).tasks ?? {}; } catch {}
  }

  // Also scan pending/active for tasks not yet in state
  for (const j of [...listJobs("pending"), ...listJobs("active")]) {
    if (j.task && !tasks[j.task]) {
      tasks[j.task] = { workflow: j.workflow, currentStep: j.step, status: "pending" };
    }
  }

  if (!Object.keys(tasks).length) {
    console.log('No tasks. Create one with: dispatch task create "description" --workflow coding-easy');
    return;
  }

  console.log("Tasks:");
  for (const [id, t] of Object.entries(tasks)) {
    console.log(`  ${id} — ${t.status} (${t.workflow}/${t.currentStep})`);
  }
}

function taskShow(taskId: string): void {
  const statePath = path.join(ROOT, "state.json");
  if (fs.existsSync(statePath)) {
    try {
      const s = JSON.parse(fs.readFileSync(statePath, "utf8"));
      if (s.tasks?.[taskId]) {
        console.log(`Task: ${taskId}`);
        console.log(JSON.stringify(s.tasks[taskId], null, 2));
      } else {
        console.log(`Task: ${taskId} (not in state)`);
      }
    } catch {}
  }

  const descPath = path.join(ROOT, "artifacts", taskId, "task.md");
  if (fs.existsSync(descPath)) {
    console.log(`\nDescription:\n  ${fs.readFileSync(descPath, "utf8").trim()}`);
  }

  console.log("\nJobs:");
  for (const folder of ["pending", "active", "done", "failed"]) {
    for (const j of listJobs(folder)) {
      if (j.task === taskId) {
        console.log(`  [${folder}] ${j.id} — step:${j.step} agent:${j.agent}`);
      }
    }
  }
}
